### ISSUE TITLE
Using /// <reference lib="dom" /> in a .d.ts typescript declaration file has global effects

### ISSUE THREAD

**Issue by pyBlob** [microsoft/TypeScript#33901]
**TypeScript Version:**  3.5.3, 3.6.3

**Search Terms:**
- triple slash lib dom
- triple slash global

**Code**

arbitrary.d.ts
```ts
/// <reference lib="dom" />
```

index.ts
```ts
export let x: Element;
```

**Expected behavior:**
Executing `npx tsc --noEmit --target esnext --lib esnext index.ts` should produce a compile error because Element is not defined nor imported and lib dom is not set globally.

The documentation for triple slash directives states the following:
https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html
- `For declaration file authors who rely on built-in types, e.g. DOM APIs or built-in JS run-time constructors like Symbol or Iterable, triple-slash-reference lib directives are recommended. Previously these .d.ts files had to add forward/duplicate declarations of such types.`
used in e.g. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/superagent/index.d.ts

**Actual behavior:**
It just compiles.

**Related Issues:**
- 
- https://github.com/microsoft/TypeScript/issues/31894
- https://github.com/microsoft/TypeScript/issues/32908 (adding skipLibChecks does not solve this problem, and should not be the right solution)

**Comment by RyanCavanaugh** [comment]
This is what a lib reference *does*.

**Comment by pyBlob** [comment]
Ah ok, so lib references also have this behaviour for all files in the compilation, not only the d.ts files. After reading the documentation another time I found the description: `For example, adding /// <reference lib="es2017.string" /> to one of the files in a compilation is equivalent to compiling with --lib es2017.string.`.

We have found this is an issue, because we had a node typescript project that uses superagent/supertest. During refactoring an import to our own "Element" method got dropped, which however didn't cause the compilation to fail. Instead it caused a seperate project to fail compilation, because types didn't match. (Local testcases would have probably caught this issue).

Is there a way to make sure that no package may change our compilation settings? (other than manually checking every file and after every update)
Is there a way to have local lib references, that are scoped to the file they are used in?

**Comment by pyBlob** [comment]
Do you think that splitting @types packages like the one for superagent into multiple packages depending on intended usage environment is a viable option?

**Comment by aleclarson** [comment]
@RyanCavanaugh Should type directives be confined to the project they're imported by? For example, let's say I have a `__stories__` folder nested in my React Native app's `src` directory, Storybook happens to depend on `@types/node`, and I have a `tsconfig.story.json` that only includes `**/__stories__` modules. The problem is that Storybook uses a type directive, and it leaks into my React Native source files, despite them never importing Storybook.

**Comment by aleclarson** [comment]
@lilred There's no workaround until the [`strictEnvironment`](https://gist.github.com/RyanCavanaugh/702ebd1ca2fc060e58e634b4e30c1c1c) proposal or something similar is implemented.
